File: how-dev-spark/src/main/scala/com/howdev/sparkdev/wordcount/WordCount1.scala
package com.howdev.sparkdev.wordcount

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

object WordCount1 {
  def main(args: Array[String]): Unit = {
    // 创建spark运行的配置信息
    // setMaster 设置运行集群(local为本地)
    // setAppName 设置job的name
    val sparkconf: SparkConf = new SparkConf().setMaster("local").setAppName("WordCountDemo")

    // 创建和Spark框架的链接
    val sc: SparkContext = new SparkContext(sparkconf)

    println("========原始数据===============")
    // 读取指定文件数据
    val txtFile: RDD[String] = sc.textFile("how-dev-spark/test-datasets/sparkdev/wordcount/word.txt")
    txtFile.foreach(println)

    println("======切割+扁平化处理==================")
    // 切割单词,扁平化处理
    val words: RDD[String] = txtFile.flatMap(
      _.split(" ")
    )
    words.foreach(println)

    println("=======分组===========================")
    // 分组
    val wordGrp: RDD[(String, Iterable[String])] = words.groupBy(word => word)
    wordGrp.foreach(println)

    println("=======统计===========================")

    def sum(pt: (String, Iterable[String])) = {
      (pt._1, pt._2.size)
    }

    //计数
    val result: RDD[(String, Int)] = wordGrp.map(sum)
    result.foreach(println)

    println("=======根据计数排序===========================")
    val sort = result.sortBy(pt => pt._2)
    sort.foreach(println)

    // 关闭连接
    sc.stop()
  }

}

File: how-dev-spark/src/main/java/com/howdev/sparkdev_java/wordcount/WordCount1.java
package com.howdev.sparkdev_java.wordcount;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Tuple2;

import java.util.Arrays;

public class WordCount1 {
    public static void main(String[] args) {
        SparkConf conf = new SparkConf();
        conf.setMaster("local[4]");
        conf.setAppName(WordCount1.class.getSimpleName());

        final JavaSparkContext jsc = new JavaSparkContext(conf);

        JavaRDD<String> rawRdd = jsc.textFile("how-dev-spark/test-datasets/sparkdev_java/wordcount.txt");

        JavaRDD<String> wordRdd = rawRdd.flatMap(line -> Arrays.asList(line.split(" ")).iterator());

        JavaPairRDD<String, Iterable<String>> wordGroupRdd = wordRdd.groupBy(word -> word);

        JavaPairRDD<String, Integer> wordCountRdd = wordGroupRdd.mapValues(iter -> {
            int count = 0;
            for (String word : iter) {
                count++;
            }
            return count;
        });

        // 按照count降序排序
        JavaPairRDD<Integer, String> countWordRdd = wordCountRdd
                .mapToPair(tuple -> new Tuple2<>(tuple._2(), tuple._1()))
                .sortByKey(false);  // false表示降序

        countWordRdd
                .collect().forEach(System.out::println);

        jsc.close();
    }
}
